# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [2]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Для того что бы это выяснить, нам надо найти и сравнить доли должников среди тех у кого есть дети и у кого их нет.

In [23]:
debtors = data.pivot_table(index='children', values ='debt', aggfunc=['count','sum','mean']) #создаем сводную таблицу, где будут отображены дети и сумма, количество и среднее количество должников
debtors.columns = ['count_debt','sum_debt','mean_debt'] # переименовываем столбцы для удобства
debtors['mean_debt'] = (debtors['mean_debt'] * 100).round(1) # переводим значения в проценты и округляем до одного знака после запятой

In [24]:
debtors.reset_index()
# выводим таблицу на экран, для красоты обнуляем индексы

children  count_debt  sum_debt  mean_debt
0         0       14091      1063        7.5
1         1        4808       444        9.2
2         2        2052       194        9.5
3         3         330        27        8.2
4         4          41         4        9.8
5         5           9         0        0.0

**Вывод:** 
Наличие детей влияет на выплату кредита в срок. Проведя анализ, мы выяснили что клиенты не имеющие детей, попадают в категорию должников в 7.5% случаев, а клиенты с 1-2 детьми в 9.2-9.5% случаях.
Скорее всего, это связано с дополнительными и непредвиденными расходами на детей.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
data['family_status'].unique() #узнаем какие значения есть в этом столбце, понимаем что их  можно разделить на 2 категории:"в браке","одиноки"

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [26]:
i = data['family_status']
def family_category(i):
    if 'женат / замужем' in i or 'гражданский брак' in i:
        return 'в браке'
    else:
        return 'одиноки'

data['family_category'] = data['family_status'].apply(family_category)


In [27]:
debtors_family = data.pivot_table(index='family_category', values ='debt', aggfunc=['count','sum','mean']) #создаем сводную таблицу, где будут отображены дети и сумма, количество и среднее количество должников
debtors_family.columns = ['count_debt','sum_debt','mean_debt'] # переименовываем столбцы для удобства
debtors_family['mean_debt'] = (debtors_family['mean_debt'] * 100).round(1)

In [28]:
debtors_family.reset_index()

family_category  count_debt  sum_debt  mean_debt
0         в браке       16395      1312        8.0
1         одиноки        4936       420        8.5

**Вывод:** Для более точных выводов мы создали таблицу из двух категорий "в браке" и "одиноки". По этой таблице мы видим, что семейное положение влияет на возврат кредита в срок. Люди состоящие в браке или сожители(я так понимаю, что здесь под гражданским браком понимается сожительство) на 0.5% чаще выплачивают кредит в срок, чем одинокие люди.
Скорее всего, это связано с тем, что в семьях кредит выплачивают оба супруга/сожителя.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

В данном случае, мы проводим те же манипуляции, что и прошлых двух заданиях, то есть создадим сводную таблицу и высчитаем среднее арифметическое. 
До этого мы разделили доходы на разные категории 'E' - самый низкий уровень дохода, 'A' - самый высокий.
0–30000 — 'E';
30001–50000 — 'D';
50001–200000 — 'C';
200001–1000000 — 'B';
1000001 и выше — 'A'.

In [29]:
data.groupby('total_income_category')['debt'].count()

total_income_category
A       25
B     5014
C    15921
D      349
E       22
Name: debt, dtype: int64

Мы видим, что в категориях А,D,Е мало информации для каких-либо выводов, поэтому создадим три категории: низкий доход, средний доход и высокий доход.

In [30]:
i = data['total_income_category']
def income_category(i):
    if 'E' in i or 'D' in i:
        return 'низкий доход'
    elif 'C' in i:
        return 'средний доход'
    else:
        return 'высокий доход'

data['income_category'] = data['total_income_category'].apply(income_category)

In [40]:
income_debtors = data.pivot_table(index='income_category', values ='debt', aggfunc=['count','sum','mean']) #создаем сводную таблицу, где будут отображены дети и сумма, количество и среднее количество должников
income_debtors.columns = ['count_debt','sum_debt','mean_debt'] # переименовываем столбцы для удобства
income_debtors['mean_debt'] = (income_debtors['mean_debt'] * 100).round(1)

In [39]:
income_debtors.reset_index()

income_category      debt
0   высокий доход  0.070649
1    низкий доход  0.061995
2   средний доход  0.084982

**Вывод:** Как мы видимо по таблице, клиенты с низким уровнем дохода реже всех становятся должниками,возможно это связано с тем, что им нечем выплачивать проценты за просрочку кредита. А клиенты со средним доходом чаще всех становятся должниками, но их и значительно больше, чем людей с низким и высоким доходом.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [62]:
goals_debtors = data.pivot_table(index='purpose_category', values ='debt', aggfunc=['count','sum','mean']) #создаем сводную таблицу, где будут отображены дети и сумма, количество и среднее количество должников
goals_debtors.columns = ['count_debt','sum_debt','mean_debt'] # переименовываем столбцы для удобства
goals_debtors['mean_debt'] = (goals_debtors['mean_debt'] * 100).round(1)

In [63]:
goals_debtors.reset_index()

purpose_category  count_debt  sum_debt  mean_debt
0    операции с автомобилем        4279       400        9.3
1  операции с недвижимостью       10751       780        7.3
2     получение образования        3988       369        9.3
3        проведение свадьбы        2313       183        7.9

**Вывод:** Чаще отдают кредит в срок те клиенты, что взяли кредит на недвижимость,скорее всего это связано  тем,что ипотеку дают клиентам с подходящим возрастом, достаточными официальным доходом и стажем работы. На втором месте те кто взял кредит на свадьбу. Должниками становятся те, кто брал кредит на образование или автомобиль, в одинаковой степени.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски могли возникнуть из-за технических ошибок, при переносе данных, например. Или в случаях,когда человек не заполнил данные о себе, в связи с чем возник пропуск.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Потому что если в данных будет переменная сильноотличающаяся от других, то она сильно повлияет на среднее значение и вывды будут неверны. Поэтому лучше заменить пропуски медианным значением

### Шаг 4: общий вывод.

Мы выполнили исследование надежности заемщиков.
Какие выводы получили:
* **Существует зависимость между количеством детей и возвратом кредита в срок**  <br/>
Если 1-2 ребенка вероятность стать должниками 9.2% - 9.5%, если 3 ребенка этот процент падает до 8.2%, для 4 детей у нас маленькая выборка, сложно делать выводы. **Самые надежные заемщики - клиенты без детей**
---
* **Семейное положение также влияет на возврат кредита в срок**<br/>
Разделив всех клиентов только на 2 группы "в браке" и "одиноки", и проанализировав их, мы можем сделать вывод, что одинокие люди на 0.5% чаще становятся должниками, то есть **клиенты в отношениях чаще отдают кредит в срок**
---
* **Уровень дохода влияет на возврат кредита в срок**<br/>
Клиенты с низким уровнем дохода не имеют возможности опатить проценты за просрочку кредита и скорее всего не совершают импульсивных покупок, поэтому должниками становятся 6.2% *на 2.3% реже чем люди со средним уровнем дохода*, клиенты с высоким уровнем дохода становятся должниками в 7.1% случаев, думаю что высокий доход связан с рисками, отсюда и просрочки. Клиенты со средним уровнем дохода самые ненадежные заемщики, поскольку во-первых их больше всего, во-вторых им чаще свойственны необдуманные покупки, а также другие факторы(потеря работы,болезнь), должниками в этом случае становятся 8.5%             **Самые надежные заемщики люди с низким и высоким доходом**
---
* **Цель кредита влияет на его возврат в срок**<br/>
Если человек берет кредит на автомобиль или образование, то вероятность того что он просрочит кредит состовляет 9.3%, скорее всего это связано с тем, что такие кредиты дают без особой проверки заемщика. Люди взявшие кредит на свадьбу становятся должниками в 7.9% случаев, а **самые надежные заемщики те, кто берет ипотеку**,для того что бы получить кредит на недвижимость надо соотвествовать некторым стандартам, иметь подходящий возраст, доход и стаж работы.
---
**Самый надежный заемщик**:
без детей, в отношениях, с низким/высоким уровнем дохода, обращается в банк за ипотекой или кредитом на свадьбу.

